### PR TITLE
feat: Support top-p and top-k

### PR DESCRIPTION
**Pull request by zhandaz**
# What does this PR do ?

This PR is a reimplementation of #1578 on the new main. It implements top-k and top-p sampling in training to ensure consistency between training and inference when sampling parameters are used.

**Status: WIP** 
1. The Megatron worker refactoring is not merged yet. So we will update later.
2. Will need to end-to-end runs for tests.

**Why is this tricky?** In Tensor Parallelism (TP), the vocabulary is sharded across GPUs (vocab-parallel). Top-k and top-p filtering require probabilities across the *full* vocabulary, not individual shards. A naive all-gather would incur large memory overhead.

**Solution:** We convert from vocab-parallel to batch-sequence-parallel layout via all-to-all communication, apply filtering on the full vocabulary, then convert back. This avoids materializing the full vocabulary on any single rank.

### Detailed Changes

- **Core sampling utilities** (`nemo_rl/models/policy/utils.py`):
  - `TrainingSamplingParams` dataclass for standardizing `top_k`, `top_p`, `temperature`.
  - `ApplyTopKTopP` autograd function with correct gradient masking.

- **Distributed logprob computation** (`nemo_rl/distributed/model_utils.py`):
  - `DistributedLogprobWithSampling` and `ChunkedDistributedLogprobWithSampling`: perform vocab-parallel to batch-sequence-parallel conversion via `all_to_all_vp2sq`/`all_to_all_sq2vp`, apply filtering, then convert back.
  - `compute_logprobs_from_logits` to centralize logprob computation across all parallel setups (TP, CP, DTensor, packed sequences).

- **Loss functions** (`nemo_rl/algorithms/loss_functions.py`):
  - `ClippedPGLossFn`, `NLLLoss`, `DPOLossFn`, `DistillationLossFn`, `PreferenceLoss` now accept `sampling_params`.
  - `ClippedPGLossFn` masks out `-inf` prev_logprobs and computes separate unfiltered logprobs for KL divergence.

- **Policy workers** (`dtensor_policy_worker.py`, `dtensor_policy_worker_v2.py`, `megatron_policy_worker.py`):
  - Initialize `self.sampling_params` from generation config and pass through to loss/logprob functions.
  - `use_reference_model()` temporarily disables top-k/top-p (retains temperature) for reference model.

- **vLLM validation** (`vllm_generation.py`):
  - Simplified: `top_k` must be `None`, `-1`, or `>= 1`; `top_p` must be `> 0`. Removed old threshold-based validation.

- **Configs & nightly**:
  - Added `megatron-sampling.yaml` (temp=0.8, top_p=0.9, top_k=50) and `megatron-temp0.6.yaml` (temp=0.6) with test scripts, registered in `nightly.txt`.

### Review Starting Points

1. **Start with the dataclass and autograd core** in `nemo_rl/models/policy/utils.py`: this defines `TrainingSamplingParams` and `ApplyTopKTopP`, which everything else builds on.
3. **Then the distributed integration** in `nemo_rl/distributed/model_utils.py`: focus on `DistributedLogprobWithSampling` and the `all_to_all_vp2sq`/`all_to_all_sq2vp` helpers to understand the TP layout conversion.
4. **Then the loss function changes** in `nemo_rl/algorithms/loss_functions.py`: particularly `ClippedPGLossFn` which has the most involved changes (handling `-inf` masking and separate KL logprobs).
5. **Policy workers** are mostly mechanical plumbing: each one follows the same pattern of init, pass-through, and reference-model save/restore.

## Tests

- **`test_sampling_params_top_k_top_p`** (`tests/unit/distributed/test_model_utils.py`): Tests `apply_top_k_top_p` forward correctness and gradient masking (gradients zero for filtered positions, preserved for non-filtered).
- **`test_sampling_params_distributed_logprob`** (`tests/unit/distributed/test_model_utils.py`): Tests `DistributedLogprobWithSampling` and `ChunkedDistributedLogprobWithSampling` in a 2-GPU TP setup, validating forward logprobs and backward gradients against a full-vocabulary baseline.
- **`test_vllm_logprobs_mode`** (`tests/unit/models/generation/test_vllm_logprobs_mode.py`): Adapted to the new `apply_top_k_top_p` return signature (logits, mask tuple).
- **Nightly convergence tests**: Two new 1N8G Megatron training runs: one with sampling (top_p=0.9, top_k=50, temp=0.8) and one with temperature-only (temp=0.6): validating reward convergence and `token_mult_prob_error` metrics.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

CC: @yuki-97 @terrykong 

## Summary by CodeRabbit

* **New Features**
  * Added experimental GRPO training configurations for Llama-3.2-1B Instruct with Megatron sampling and temperature control variants.
  * Enabled top-k/top-p token filtering during distributed training with improved gradient handling for sampling-based approaches.

* **Bug Fixes**
  * Simplified parameter validation for generation settings with clearer error messaging.

* **Tests**
  * Added comprehensive test coverage for distributed sampling operations and token filtering behavior.

### CODE REVIEW

**Comment by terrykong** [comment]
closing since this has been merged into main in another PR
